src/xunit.analyzers/InlineDataShouldBeUniqueWithinTheory.cs

```diff
@@ -0,0 +1,154 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Xunit.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class InlineDataShouldBeUniqueWithinTheory : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            Descriptors.X1025_InlineDataShouldBeUniqueWithinTheory);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                var compilationTypes = new CompilationTypes(compilationStartContext.Compilation);
+
+                if (!compilationTypes.IsCompilationCapableOfBeingProcessedByThisAnalyzer)
+                    return;
+
+                compilationStartContext.RegisterSymbolAction(
+                    symbolContext => AnalyzeMethod(symbolContext, compilationTypes), SymbolKind.Method);
+            });
+        }
+
+        private static void AnalyzeMethod(SymbolAnalysisContext context, CompilationTypes compilationTypes)
+        {
+            var method = (IMethodSymbol) context.Symbol;
+
+            var methodAllAttributes = method.GetAttributes();
+            if (!methodAllAttributes.ContainsAttributeType(compilationTypes.Theory))
+                return;
+
+            var wellFormedInlineDataAttributes = methodAllAttributes
+                .Where(a => a.AttributeClass == compilationTypes.InlineData
+                            && HasAttributeDeclarationNoCompilationErrors(a, compilationTypes));
+
+            AnalyzeInlineDataAttributesWithinTheory(context, wellFormedInlineDataAttributes);
+        }
+
+        private static void AnalyzeInlineDataAttributesWithinTheory(SymbolAnalysisContext context,
+            IEnumerable<AttributeData> inlineDataAttributes)
+        {
+            var uniquenessMapByHashCodes = new Dictionary<int, AttributeData>();
+
+            foreach (var currentInlineData in inlineDataAttributes)
+            {
+                context.CancellationToken.ThrowIfCancellationRequested();
+
+                IEnumerable<object> arguments = GetTestArguments(context, currentInlineData);
+                int argumentsHash = GetTestArgumentsAggregatedHashCode(arguments);
+
+                if (uniquenessMapByHashCodes.ContainsKey(argumentsHash))
```

Comment:
Just testing the hashes for equality is insufficient and could lead to false positives. You also need to test the two argument collections and see if they are "equal". This could be done by using a Set<AttributeData> with a custom comparer.

(_Technically_, the argument equality check should also consider the corresponding target parameter's type conversion, e.g. `[InlineData(1), InlineData(1L), Theory] void Test(long p) {}` could be marked as having a duplicate even though the arguments themselves are not strictly the same but the test method ends up being called with the same value. **However**, this is probably overkill as dealing with conversions is tricky and it gets further complicated by generic theory methods)